addon/addon/models/sprite-snapshot-node-builder.ts

```diff
@@ -0,0 +1,344 @@
+  let spriteLast = spriteModifier.lastBounds;
+  let contextCurrent = animationContext.currentBounds;
+  let contextLast = animationContext.lastBounds;
+  if (spriteCurrent && spriteLast && contextCurrent && contextLast) {
+    let parentLeftChange = contextCurrent.left - contextLast.left;
+    let parentTopChange = contextCurrent.top - contextLast.top;
+
+    return (
+      spriteCurrent.left - spriteLast.left - parentLeftChange !== 0 ||
+      spriteCurrent.top - spriteLast.top - parentTopChange !== 0 ||
+      spriteCurrent.width - spriteLast.width !== 0 ||
+      spriteCurrent.height - spriteLast.height !== 0
+    );
+  }
+  return true;
+}
+
+export function filterToContext(
+  spriteTree: SpriteTree,
+  animationContext: AnimationContext,
+  spriteModifiers: Set<SpriteModifier>,
+  opts: GetDescendantNodesOptions = { includeFreshlyRemoved: false }
+): Set<SpriteModifier> {
+  let contextDescendants = spriteTree.descendantsOf(animationContext, {
+    ...opts,
+    filter(childNode: SpriteTreeNode) {
+      return !(
+        childNode.isContext &&
+        (childNode.contextModel as AnimationContext).isStable
+      );
+    },
+  });
+  return new Set(
+    [...spriteModifiers].filter((m) => contextDescendants.includes(m))
+  );
+}
+
+export class SpriteSnapshotNode {
+  controllingContext: AnimationContext;
+  insertedSprites: Set<Sprite> = new Set();
+  removedSprites: Set<Sprite> = new Set();
+  keptSprites: Set<Sprite> = new Set();
+
+  constructor(context: AnimationContext) {
+    this.controllingContext = context;
+  }
+
+  addSprite(
+    sprite: Sprite,
+    spriteModifier: SpriteModifier,
+    context: AnimationContext,
+    counterpartModifier?: SpriteModifier
+  ) {
+    if (sprite.type === SpriteType.Kept) {
+      assert(
+        'kept sprite should have lastBounds and currentBounds',
+        spriteModifier.lastBounds &&
+          context.lastBounds &&
+          spriteModifier.currentBounds &&
+          context.currentBounds
+      );
+
+      sprite.initialBounds = new ContextAwareBounds({
+        element: spriteModifier.lastBounds,
+        contextElement: context.lastBounds,
+      });
+      sprite.finalBounds = new ContextAwareBounds({
+        element: spriteModifier.currentBounds,
+        contextElement: context.currentBounds,
+      });
+      sprite.initialComputedStyle = spriteModifier.lastComputedStyle;
+      sprite.finalComputedStyle = spriteModifier.currentComputedStyle;
+
+      if (sprite.counterpart) {
+        assert(
+          'counterpart modifier should have been passed',
+          counterpartModifier
+        );
+
+        assert(
+          'kept sprite counterpart should have lastBounds and currentBounds',
+          counterpartModifier.lastBounds && counterpartModifier.currentBounds
+        );
+
+        // TODO: double check if we're setting the correct bounds here
+        if (counterpartModifier) {
+          sprite.counterpart.initialBounds = new ContextAwareBounds({
```

Comment:
If it's an inserted sprite converted into a kept sprite, we need to set the sprite's initial bounds to the counterpart's initial bounds too.

Reply:
I've got a bunch of modifications for this part locally too as I've now mostly integrated an updated version of the intermediate sprite system. We now also only create those for sprites that are actually animating which was not the case before.

Reply:
We resolved this in the last commit.